src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -218,26 +219,30 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
     for (Commands::const_iterator i = commands.begin(); i != commands.end();
 	 ++i) {
       (void)admin_socket->unregister_command(i->first);
-      delete i->second;
     }
     commands.clear();
   }
 
   bool call(std::string command, cmdmap_t& cmdmap, std::string format,
 	    bufferlist& out) override {
-    Mutex::Locker locker(lock);
-    Commands::const_iterator i = commands.find(command);
-    assert(i != commands.end());
+    CommandRef command_ref;
+    {
+      Mutex::Locker locker(lock);
+      Commands::const_iterator i = commands.find(command);
+      assert(i != commands.end());
+      command_ref = i->second;
+    }
     Formatter *f = Formatter::create(format);
     stringstream ss;
-    bool r = i->second->call(f, &ss);
+    bool r = command_ref->call(f, &ss);
```

Comment:
@dillaman I think the scenario you described is not possible. Because on image replayer destruction it need to unregister the hook commands first. And it will block until an inflight command is completed. For this reason it looks like using a shared pointer is not necessary (though we still need to hold the lock to keep Commands structure stable when looking for the command pointer). But it reveals another issue: it looks we are not allowed to unregister hook commands (call `~ImageReplayerAdminSocketHook` destructor) when holding ImageReplayer lock, as it may lead to a deadlock.